File: c/subprocess.c
#include <errno.h>
#include <stdlib.h>
#include <sys/ioctl.h>
#include <unistd.h>

int sp_forkexec(char **argv,
        pid_t *child_pid,
        int cldin,
        int cldout,
        int clderr,
        int ptm)
{
    int rc;

    *child_pid = fork();
    if (*child_pid == 0)
    {
        close(ptm);

        if (cldin >= 0)
        {
            rc = dup2(cldin, STDIN_FILENO);
            if (rc < 0)
                goto child_fail;
            if (ptm < 0)
                close(cldin);
        }

        if (cldout >= 0)
        {
            rc = dup2(cldout, STDOUT_FILENO);
            if (rc < 0)
                goto child_fail;
            // stdout and stderr may be the same, but stdin shouldn't be unless it's a term
            if (ptm < 0 && cldout != clderr)
                close(cldout);
        }

        if (clderr >= 0)
        {
            rc = dup2(clderr, STDERR_FILENO);
            if (rc < 0)
                goto child_fail;
            if (ptm < 0)
                close(clderr);
        }

        // If we're attaching to a PTY, extra work is required before we exec away
        if (ptm >= 0)
        {
            close(cldin);

            setsid();
            rc = ioctl(STDIN_FILENO, TIOCSCTTY, 1);
            if (rc < 0)
                goto child_fail;
        }

        execvp(argv[0], argv);

child_fail:
        exit(EXIT_FAILURE);
    }
    else if (*child_pid > 0)
    {
        if (ptm >= 0)
            close(cldin);
        return 0;
    }

    // Fork failed
    return errno;
}

File: src/lib/subprocess.rs
use std::ffi::CString;
use std::ffi::OsStr;
use std::os::unix::ffi::OsStrExt;

extern crate libc;

use super::pty;

#[repr(C)]
#[allow(dead_code)]
enum idtype_t {
    P_ALL,
    P_PID,
    P_PGID,
}

#[repr(C)]
#[allow(dead_code)]
enum si_code {
    CLD_NONE,
    CLD_EXITED,    // Child has exited.
    CLD_KILLED,    // Child was killed.
    CLD_DUMPED,    // Child terminated abnormally.
    CLD_TRAPPED,   // Traced child has trapped.
    CLD_STOPPED,   // Child has stopped.
    CLD_CONTINUED, // Stopped child has continued.
}


#[repr(C)]
struct siginfo_t {
    si_signo: libc::c_int,
    si_errno: libc::c_int,
    si_code: si_code,
    _pad: [libc::c_int; 29],
}

extern {
    fn waitid(idtype: idtype_t,
              id: libc::id_t,
              infop: *mut siginfo_t,
              options: libc::c_int);

    pub fn sp_forkexec(argv: *const *const libc::c_char,
                    child_pid: *mut libc::pid_t,
                    cldin: libc::c_int,
                    cldout: libc::c_int,
                    clderr: libc::c_int,
                    ptm: libc::c_int) -> libc::c_int;
}

pub struct Child {
    pid: libc::pid_t,
}

pub struct Command {
    args: Vec<CString>,
    argv: Vec<*const libc::c_char>,
    stdin: libc::c_int,
    stdout: libc::c_int,
    stderr: libc::c_int,
    ptm: libc::c_int,
}

pub enum Status {
    Running,
    Exited(libc::c_int),
    Killed,
}

impl Child {
    pub fn wait_noblock(&self) -> Status {
        let mut status = siginfo_t {
            si_signo: 0,
            si_errno: 0,
            si_code: si_code::CLD_NONE,
            _pad: [0; 29],
        };
        unsafe {
            waitid(idtype_t::P_PID,
                   self.pid as libc::id_t,
                   &mut status,
                   libc::WNOHANG | libc::WEXITED);
        }
        if status.si_signo == 0 {
            Status::Running
        } else {
            match status.si_code {
                si_code::CLD_EXITED => Status::Exited(status.si_errno),
                si_code::CLD_KILLED => Status::Killed,
                si_code::CLD_DUMPED => Status::Killed,
                _ => Status::Running,
            }
        }
    }
}

impl Command {
    pub fn new() -> Command {
        Command {
            args: Vec::new(),
            argv: Vec::new(),
            stdin: -1,
            stdout: -1,
            stderr: -1,
            ptm: -1,
        }
    }

    pub fn arg<S: AsRef<OsStr>>(&mut self, arg: S) -> &mut Command {
        let arg = CString::new(arg.as_ref().as_bytes()).unwrap();
        self.argv.push(arg.as_ptr());
        self.args.push(arg);
        self
    }

    pub fn stdio_from_pty(&mut self, term: &mut pty::Pty) -> &mut Command {
        let (ptm, pts) = term.get_fds();
        self.stdin = pts;
        self.stdout = pts;
        self.stderr = pts;
        self.ptm = ptm;
        term.disown_pts();
        self
    }

    pub fn run(&mut self) -> Result<Child, libc::c_int> {
        self.argv.push(0 as *const _);

        let mut ret = Child { pid: -1 };
        let rc = unsafe {
            sp_forkexec(self.argv.as_ptr(),
                                 &mut ret.pid,
                                 self.stdin,
                                 self.stdout,
                                 self.stderr,
                                 self.ptm)
        };

        match rc {
            0 => Ok(ret),
            e => Err(e),
        }
    }
}
